### PR TITLE
[Small Feature] Spatial Searching Balanced Kd Tree

### PR DESCRIPTION
**Pull request by danston**
This draft PR tries to modify the Kd tree class in order to support a new balanced Kd tree build from [this paper](http://jcgt.org/published/0004/01/03/).

## Release Management

* Affected package(s): `Spatial_searching`
* Issue(s) solved (if any): #5812 
* Feature/Small Feature (if any): Small Feature [not yet submitted]
* Link to compiled documentation (obligatory for small feature): docs [still to be added]
* License and copyright ownership: GF

Tasks:

- [ ] Add docs.
- [ ] Submit a small feature.
- [ ] Finish implementation.
- [ ] Test for various edge cases.
- [ ] Update changes.md.
- [ ] Check for memory leaks.
- [ ] Remove all experimental code (such as preprocess in Kd tree and modifs in AABB tree).

### CODE REVIEW

File: Spatial_searching/include/CGAL/Kd_tree.h [hunk 1]
```diff
@@ -188,14 +188,20 @@ class Kd_tree {
     split(sep, c, c_low);
     nh->set_separator(sep);
 
-    handle_extended_node (nh, c, c_low, UseExtendedNode());
+    if (!std::is_same<Splitter, CGAL::Balanced_splitter<SearchTraits>>::value) {
```
[the remark this answers is not in the document]
Comment (danston) [reply Spatial_searching/include/CGAL/Kd_tree.h]: I moved `is_same` to the static global field.

File: Spatial_searching/include/CGAL/Kd_tree.h [hunk 2]
```diff
@@ -188,14 +188,20 @@ class Kd_tree {
...
+      handle_extended_node (nh, c, c_low, UseExtendedNode());
+    }
 
     if (try_parallel_internal_node_creation (nh, c, c_low, tag))
       return;
 
+    std::cout << "c low size: " << c_low.size() << std::endl;
```
[the remark this answers is not in the document]
Comment (danston) [reply Spatial_searching/include/CGAL/Kd_tree.h]: I totally agree. Once the code is tested and I see that everything is correct, I will hide them. For the moment, I commented out all my internal `couts`.

[the remark this answers is not in the document]
Comment (danston) [reply Spatial_searching/include/CGAL/Kd_tree.h]: The thing is that while testing I have a lot of them and I constantly add/remove some so putting them behind ifdefs each time I add/remove makes the code very noisy imho.

File: Spatial_searching/include/CGAL/Kd_tree.h [hunk 3]
```diff
@@ -288,6 +295,20 @@ class Kd_tree {
     return pts.empty();
   }
 
+  void preprocess(const FT distance_threshold = FT(1) / FT(1000)) {
```
[the remark this answers is not in the document]
Comment (danston) [reply Spatial_searching/include/CGAL/Kd_tree.h]: Slightly fixed. I will later remove it completely. I added this in the TODO.

File: Spatial_searching/include/CGAL/Kd_tree.h [hunk 4]
```diff
@@ -336,12 +357,21 @@ class Kd_tree {
                                "Parallel_tag is enabled but TBB is unavailable.");
 #endif
 
-    Point_container c(dim_, data.begin(), data.end(),traits_);
+    bool create_balanced_tree = false;
```
[the remark this answers is not in the document]
Comment (danston) [reply Spatial_searching/include/CGAL/Kd_tree.h]: Done.

File: Spatial_searching/include/CGAL/Kd_tree.h [hunk 5]
```diff
@@ -336,7 +448,50 @@ class Kd_tree {
                                "Parallel_tag is enabled but TBB is unavailable.");
 #endif
 
-    Point_container c(dim_, data.begin(), data.end(),traits_);
+    std::vector< std::vector<FTP> > references;
+    long start = -1, end = -1;
+    if (m_create_balanced_tree) {
+
+      references.resize(dim_, std::vector<FTP>(data.size()));
+      std::vector<std::size_t> ref_end(references.size());
```
Comment (afabri) [inline Spatial_searching/include/CGAL/Kd_tree.h:456]: `references,size()` is `dim_`.

File: Spatial_searching/include/CGAL/Point_container.h [hunk 6]
```diff
@@ -55,6 +56,97 @@ class Point_container {
...
+
+      CGAL_assertion(p != FTP());
+      CGAL_assertion(q != FTP());
+      FT coords_diff = FT(0);
+      for (std::size_t k = 0; k < dim; ++k) {
+        const std::size_t idx = (k + axis) % dim; // stay inside dim
+        coords_diff = p[idx] - q[idx];
+        if (coords_diff != FT(0)) break;
+      }
+      return coords_diff;
```
Comment (afabri) [inline Spatial_searching/include/CGAL/Point_container.h:91]: Should the result not be of type `bool`  with `true` for the first parameter being larger than the second?

File: Spatial_searching/include/CGAL/Point_container.h [hunk 7]
```diff
@@ -330,6 +422,32 @@ class Point_container {
     return b;
   }
 
+  template <class Separator>
+  void balanced_split(Point_container<Traits>& c, Separator& sep)
```
Comment (afabri) [inline Spatial_searching/include/CGAL/Point_container.h:426]: Why is `sep` not `const`?

File: Spatial_searching/include/CGAL/Splitters.h [hunk 8]
```diff
@@ -303,5 +303,179 @@ namespace CGAL {
...
+      for (std::size_t k = 1; k < dim; ++k) {
+        lower = start - 1;
+        upper = median;
+
+        for (long i = start; i <= end; ++i) { // always skip median key
+          const FT compare = compare_keys(
+            references[k][i], references[0][median], axis, dim);
+          if (compare < FT(0)) {
+            references[k-1][++lower] = references[k][i];
+          } else if (compare > FT(0)) {
```
Comment (afabri) [inline Spatial_searching/include/CGAL/Splitters.h:412]: I thought it cannot be zero.

File: Spatial_searching/include/CGAL/Kd_tree.h (hunk 5 above)
Comment (danston) [reply Spatial_searching/include/CGAL/Kd_tree.h]: This part has changed after we did a recent speedup.

File: Spatial_searching/include/CGAL/Point_container.h (hunk 6 above)
Comment (danston) [reply Spatial_searching/include/CGAL/Point_container.h:84]: Not really, because we need to know also when the super keys are equal that is `coords_diff = 0`.

File: Spatial_searching/include/CGAL/Point_container.h (hunk 7 above)
Comment (danston) [reply Spatial_searching/include/CGAL/Point_container.h:398]: Because we update this splitter inside this method.

File: Spatial_searching/include/CGAL/Splitters.h (hunk 8 above)
Comment (danston) [reply Spatial_searching/include/CGAL/Splitters.h:427]: It can.

Comment (afabri) [reply Spatial_searching/include/CGAL/Splitters.h:427]: And  do you use `compare` as a `FT`  other than for comparing with zero?  If not use ` Comparison_result`.  And don't perform an `operator-()`.  And in case you really have to use a `FT`, then call functions such as [` is_zero()`](https://doc.cgal.org/latest/Algebraic_foundations/group__PkgAlgebraicFoundationsRef.html#ga997aad1528e7e744e54a807caa2be56e)

File: Spatial_searching/include/CGAL/Point_container.h (hunk 7 above)
Comment (afabri) [reply Spatial_searching/include/CGAL/Point_container.h:398]: Where?

Comment (danston) [reply Spatial_searching/include/CGAL/Point_container.h:398]: Ooo, You are completely right. I had a look at the old version where the `balanced_split()` was a part of the `Splitters.h` and I have updated the `sep` inside. This one does not require it. I will fix it!

File: Spatial_searching/include/CGAL/Splitters.h (hunk 8 above)
Comment (danston) [reply Spatial_searching/include/CGAL/Splitters.h]: Done.
